### ISSUE TITLE
Transmeta Crusoe?

### ISSUE THREAD

**Issue by piranna** [losfair/metamachine#1]
Isn't this project somewhat similar to what Transmeta Crusoe microprocessors tried to achieve?

By the way, I'm currently working on a [WebAssembly implementation in FPGAs](https://github.com/piranna/wasmachine).

**Comment by piranna** [comment]
https://en.wikipedia.org/wiki/Transmeta_Crusoe

**Comment by losfair** [comment]
Yeah this project is similar to Transmeta Crusoe in that they both try to emulate another instruction set with low-level software in a different hardware architecture. Actually I think a lot of modern processors (like those from Intel or AMD) use similar approaches as well, except that how the translation works internally is completely transparent to software (maybe microcode update is an exception).

However, it seems that Transmeta Crusoe with its translation layer (CMS) and this project (Metamachine) try to achieve a similar goal in different ways. CMS is a JIT compiler that dynamically compiles x86 code to a proprietary VLIW instruction set, while Metamachine tries to do some form of "meta-execution" statically and purely at hardware level, by having a small piece of microcode-like program, which resides in hardware registers, "interpreting" a higher level instruction set. IMO these two different approaches both have advantages and disadvantages, but a complex x86 JIT compiler as the foundation of every other application sounds a bit unsafe.

Aside from that, this project's goal is to provide a modern, open-source, 64-bit implementation of a software-defined instruction set architecture.

Btw your work on the hardware implementation of WebAssembly is awesome :)

**Comment by piranna** [comment]
> Actually I think a lot of modern processors (like those from Intel or AMD) use similar approaches as well, except that how the translation works internally is completely transparent to software (maybe microcode update is an exception).

Yes, Intel CPUs are internally RISC, but offer a CISC ISA for retrocompatibility, and microcode convert them on the fly. That, and the fact their internal RISC implementation is not directly accesible anyway from user-space code...


> Btw your work on the hardware implementation of WebAssembly is awesome :)

Yeah, unluckily I don't have almost time nor experience with FPGAs to be able to carry it to anywhere, at least in a short term place... I had the idea 4 years ago, and still doesn't have anything usable... :-/
